### PR TITLE
🕸️ Implement cluster commands

### LINKED ISSUE #37: Cluster API
- https://redis.io/commands#cluster

### PR DESCRIPTION
**Pull request by tumile**
# :construction: WIP: Cluster API

This changeset will resolve #37 :rocket:

## Commands to implement and test

- [x] cluster_addslots
- [ ] ~~cluster_bumpepoch~~
- [x] cluster_countfailurereports
- [x] cluster_countkeysinslot
- [x] cluster_delslots
- [x] cluster_failover
- [x] cluster_flushslots
- [x] cluster_forget
- [x] cluster_getkeysinslot
- [x] cluster_info
- [x] cluster_keyslot
- [x] cluster_meet
- [x] cluster_myid
- [x] cluster_nodes
- [x] cluster_replicate
- [x] cluster_reset
- [x] cluster_saveconfig
- [ ] ~~cluster_setconfigepoch~~
- [x] cluster_setslot
- [x] cluster_slaves
- [x] cluster_replicas
- [x] cluster_slots
- [x] readonly
- [x] readwrite

## Additional requirements

- [x] Update README
- [x] Match coding style and argument names

### CODE REVIEW

**Comment by tumile** [comment]
I didn't implement `CLUSTER BUMPEPOCH` and `CLUSTER SET-CONFIG-EPOCH` because per the Redis docs, config epoch management should be performed internally by the cluster. Looking at [jedis](https://github.com/xetorthio/jedis/blob/master/src/main/java/redis/clients/jedis/commands/ClusterCommands.java) and [go-redis](https://github.com/go-redis/redis/blob/dc52593c8c63bc2a05796ec44efd317fd3e14b64/commands.go#L297-L315), they left it out as well. Let me know if you think they should be in the API.

**Comment by tumile** [comment]
The cluster tests need to be run against a cluster enabled redis-server, which will interfere with other tests since in cluster mode some functionalities are disabled. So for now I don't include them in `redis_test.ts`, but they can be run separately. As I investigate go-redis, they spin up standalone redis-servers for testing with suitable mode. @uki00a What is the best way to move forward with the tests here?

**Comment by uki00a** [comment]
First of all, thanks for submitting this PR @tumile!

> they spin up standalone redis-servers for testing with suitable mode.

That is exactly what I want to do. (#104)
Evetually, I want to run the cluster tests in the CI :smile:

**Comment by tumile** [comment]
I see. Has there been any plan for #104? This PR can wait till then don't you think?

**Comment by uki00a** [comment]
@tumile 
> Has there been any plan for #104?

I added `startRedisServer()` function in `tests/test_util.ts`. (#113)
With it, could you include the cluster tests in CI?

**Comment by tumile** [comment]
I'm thinking about adding a `startRedisCluster` or something to, well, start multiple servers for the tests. I'll update as soon as I get it done.

**Comment by tumile** [comment]
I currently use a timeout at the end of file to clean up the redis cluster after tests in `cluster_test.ts`. There's no `afterAll` hook or something like that in Deno I believe. This is a problem when running with other long running tests like `pubsub_test.ts` or `bzpopmin timeout`, the cluster may be cleaned up before the cluster tests have a chance to run. My workaround now is to set the cleanup timeout really high (20s) to make sure every tests have run. This is unsound.

https://github.com/tumile/deno-redis/blob/6942c252a67ecaaf74c3280c7300a0577cdef802/tests/cluster_test.ts#L108-L112

Is there a way to ensure all tests have run before cleanup? I tried putting cleanup in another `test`, but Deno complains about leaking.

**Comment by Terkwood** [comment]
https://deno.land/manual/testing

You may have looked at this already, and it's my first time reading it, but I don't see any simple/prepackaged way to accomplish what you're asking. Will probably require a creative hack or some third party test util(if such a thing exists)?

**Comment by tumile** [comment]
There is an option to run each test file in serial, so test timeouts will not interfere with others. I don't think this is a bad idea, with #114 it's probably good to put all commands tests in a single file and start a redis instance. Tests like cluster and pubsub tests, which require customized instances, will be run separately.

Still, this doesn't solve using timeout for cleanup, but a test will not suddenly fail if some others take longer than expected.

**Comment by uki00a** [comment]
> Is there a way to ensure all tests have run before cleanup? I tried putting cleanup in another test, but Deno complains about leaking.

How about the following? :thinking:
```typescript
async function makeClusterTest(): Promise<{
  test: (name: string, fn: () => Promise<void>) => void,
  client: Redis
}> {
  const cleanupCluster = await startRedisCluster();
  const client = await connect({ hostname: "127.0.0.1", port: 7000 });
  const tests = [] as Array<{ name: string, fn: () => Promise<void> }>;

  function test(name: string, fn: () => Promise<void>): void {
    tests.push({ name, fn });
  }

  function runTests(): Promise<void> {
    const promises = [] as Promise<void>[];
    for (const test of tests) {
      Deno.test(test.name, () => {
        const promise = test.fn();
        promises.push(promise);
        return promise;
      });
    }
    return Promise.allSettled(promises).finally(() => {
      cleanupCluster();
      client.close();
    });
  }

  return { test, client, runTests };
}
```

```typescript
const { test, client, runTests } = makeClusterTest();

test("[cluster] flushslots", async () => {
  ...
});

...

runTests();
```

**Comment by tumile** [comment]
Looks promising, I'll test it in a bit. How about wrapping the makeTest/makeClusterTest in a class? Probably cleaner.

**Comment by tumile** [comment]
@uki00a can you review my test implementation? I think if it's good, we can roll this out to other tests as well!

File: tests/test_util.ts [hunk 1]
```diff
@@ -1,6 +1,105 @@
...
+export class TestSuite {
+  private encoder = new TextEncoder();
+
+  private tests: { name: string; func: TestFunc }[] = [];
+  private beforeEachs: TestFunc[] = [];
+  private afterAlls: TestFunc[] = [];
+
+  constructor(private prefix: string) {}
+
+  startRedis = async ({
```
Comment (uki00a) [inline tests/test_util.ts:20]: I think `TestSuite` should not have methods related to Redis.
Is it possible to change this method to a normal function?

File: tests/test_util.ts [hunk 2]
```diff
@@ -1,6 +1,105 @@
...
+      cmd: ["redis-server", `testdata/${port}/redis.conf`],
+      stdin: "null",
+      stdout: "null",
+    });
+
+    // Ample time for server to finish startup
+    await delay(500);
+    return { path, process };
+  };
+
```
Comment (uki00a) [inline tests/test_util.ts:48]: Ditto

File: tests/test_util.ts [hunk 3]
```diff
@@ -1,6 +1,105 @@
...
+    });
+
+    try {
+      await Promise.allSettled(promises);
+    } finally {
+      this.afterAlls.forEach(async (f) => await f());
+    }
+  };
+
+  teardown(server: TestServer): void {
```
Comment (uki00a) [inline tests/test_util.ts:85]: Ditto

File: tests/test_util.ts [hunk 4]
```diff
@@ -1,6 +1,105 @@
...
+    Deno.removeSync(server.path, { recursive: true });
+    server.process.close();
+  }
+}
+
+async function exists(filename: string): Promise<boolean> {
```
Comment (uki00a) [inline tests/test_util.ts:91]: Is this function adopted from `std/fs/exists.ts`?
If so, please add the copyright header.
https://deno.land/manual/contributing/style_guide

Comment (tumile) [reply tests/test_util.ts:91]: uh no I wrote it. But now seeing there is something like this should I use the std version?

**Comment by tumile** [comment]
I've addressed the comments 😁. Please review again

**Comment by uki00a** [comment]
Released in [v0.11.0](https://github.com/denolib/deno-redis/releases/tag/v0.11.0)!

**Pull request merged**